core/node/rpc/node2nodeauth/cert.go

```diff
@@ -0,0 +1,208 @@
+	// The given address must correspond to the signature creator address.
+	Address string `asn1:"utf8"`
+	// Signature is the signature of the cert's public key.
+	Signature []byte `asn1:"octet"`
+}
+
+// verifyCert verifies the node-2-node client certificate.
+// The certificate must have a custom node2nodeCertExt extension.
+func verifyCert(logger *zap.SugaredLogger, nodeRegistry nodes.NodeRegistry, cert *x509.Certificate) error {
+	if cert == nil {
+		return RiverError(Err_UNAUTHENTICATED, "No node-2-node client certificate provided").LogError(logger)
+	}
+
+	if cert.Subject.CommonName != node2NodeCertName {
+		return RiverError(Err_UNAUTHENTICATED, "Unexpected node-2-node client certificate name").LogError(logger)
+	}
+
+	now := time.Now()
+	if cert.NotBefore.After(now) || cert.NotAfter.Before(now) {
+		return RiverError(Err_UNAUTHENTICATED, "Certificate expired").LogError(logger)
+	}
+
+	var certExt node2NodeCertExt
+	var found bool
+	for _, ext := range cert.Extensions {
+		if ext.Id.Equal(node2NodeCertExtOID) {
+			if _, err := asn1.Unmarshal(ext.Value, &certExt); err != nil {
+				return AsRiverError(err, Err_UNAUTHENTICATED).Message("Failed to unmarshal extra extension").LogError(logger)
+			}
+			found = true
+			break
+		}
+	}
+	if !found {
+		return RiverError(Err_UNAUTHENTICATED, "Node-2-node cert extension not found").LogError(logger)
+	}
+
+	publicKeyDER, err := x509.MarshalPKIXPublicKey(cert.PublicKey)
+	if err != nil {
+		return AsRiverError(err, Err_UNAUTHENTICATED).Message("Failed to marshal public key").LogError(logger)
+	}
+
+	sigPublicKey, err := ethcrypto.SigToPub(
+		crypto.TownsHashForCert.Hash(publicKeyDER).Bytes(),
+		certExt.Signature,
+	)
+	if err != nil {
+		return AsRiverError(err, Err_UNAUTHENTICATED).Message("Failed to extract public key from the cert signature").LogError(logger)
+	}
+	recoveredAddr := ethcrypto.PubkeyToAddress(*sigPublicKey)
+
+	if recoveredAddr.Cmp(common.HexToAddress(certExt.Address)) != 0 {
+		return RiverError(Err_UNAUTHENTICATED, "Node-2-node cert signature does not match address").LogError(logger)
+	}
+
+	node, err := nodeRegistry.GetNode(recoveredAddr)
+	if err != nil {
+		return err
+	}
+
+	// Check that that sender node is operational
+	if node.Status() != river.NodeStatus_Operational {
+		return RiverError(Err_UNAUTHENTICATED, "Node is not operational").LogError(logger)
+	}
+
+	return nil
+}
+
+// CertGetter returns a GetClientCertFunc that provides a node-2-node client certificate.
+func CertGetter(logger *zap.SugaredLogger, wallet *crypto.Wallet) http_client.GetClientCertFunc {
+	const (
+		certTTL      = time.Hour * 24
+		certRenewGap = time.Minute
+	)
+
+	var (
+		lock sync.Mutex
+		cert *tls.Certificate
+		exp  time.Time
+	)
+
+	return func(info *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+		lock.Lock()
+		defer lock.Unlock()
+
+		if cert == nil || time.Now().Add(certRenewGap).After(exp) {
+			newCert, err := createCert(logger, wallet)
```

Comment:
`createCert` creates the certificate with:

```
template := &x509.Certificate{
		...
		NotBefore:       time.Now(),
		NotAfter:        time.Now().Add(time.Hour),
		...
	}
```

That means the certificate expires in 1 hour?

Later in this function: `exp = time.Now().Add(certTTL)` with `certTTL = time.Hour * 24`

And `exp` is used to determine if a new certificate must be generated. Does this mean that when the first time a certificate is created it expires after 1 hour and this function only generates a new one after 24 hours (or a restart?)